ESA Hubble metadata search improvements

jespinosaar opened the pull request:
New method query_by_criteria, removing handler to download files and metadata and _request with parameters

jespinosaar commented:
Dear Astroquery Team,
It seems that the code style fails in file we have not changed in this PR, what should we do? Thanks!

keflavich commented on astroquery/esa/hubble/core.py:
```diff
@@ -111,22 +75,21 @@ def download_product(self, observation_id, calibration_level="RAW",
-            if verbose:
-                log.info("Wrote {0} to {1}".format(link, filename))
-            return filename
+        log.info(self.copying_string.format(filename))
```
This should still be under `if verbose`

jespinosaar replied:
Implemented

keflavich commented on astroquery/esa/hubble/core.py:
```diff
@@ -189,18 +154,25 @@ def get_postcard(self, observation_id, calibration_level="RAW",
-        self._handler.get_file(filename, response=result, verbose=verbose)
+            filename = observation_id
+
+        log.info(self.copying_string.format(filename))
```
same here (`if verbose`)

keflavich commented on astroquery/esa/hubble/core.py:
```diff
@@ -233,18 +205,17 @@ def cone_search(self, coordinates, radius=0.0, filename=None,
-                                       timeout=self.TIMEOUT)
+        response = self._request('GET',
+                                 self.metadata_url,
+                                 params=payload)
```
were the `cache=` and `timeout=` keywords dropped here intentionally?  I suspect at least `TIMEOUT` was not

keflavich commented on docs/esa/hubble.rst:
```diff
@@ -89,8 +143,147 @@ It will also download a file storing all metadata for all observations
+ + HLA
+ + HST
+ 
+ Do not forget that this is a list of elements, so it must be defined as ['HST'], ['HST', 'HLA']...
```
In this case, the documentation above should be `list of strings, optional`, since a string won't be accepted

keflavich commented on docs/esa/hubble.rst:
```diff
@@ -89,8 +143,147 @@ It will also download a file storing all metadata for all observations
+ 
+ Do not forget that this is a list of elements, so it must be defined as ['HST'], ['HST', 'HLA']...
+ 
+- **instrument_name** (*str, optional*): Name/s of the instrument/s used to generate the dataset. This is also a list of elements.
```
style suggestion: use `Name(s)` instead of `Name/s`, etc.

keflavich commented on docs/esa/hubble.rst:
```diff
@@ -89,8 +143,147 @@ It will also download a file storing all metadata for all observations
+  Table length=12965
+  ...
+
+As it has been mentioned, these parameters are optional and it is not necessary to define all of them to execute this function, as the search will be adapted to the ones that are available.
```
```suggestion
As has been mentioned, these parameters are optional and it is not necessary to define all of them to execute this function, as the search will be adapted to the ones that are available.
```

keflavich commented on docs/esa/hubble.rst:
```diff
@@ -89,8 +143,147 @@ It will also download a file storing all metadata for all observations
+                                            async_job = False, 
+                                            output_file = 'output5.vot.gz')
+ 
+If none of them are defined, this function will retrieve all the observations.
```
```suggestion
If no criteria are specified to limit the selection, this function will retrieve all the observations.
```

Does "all the observations" mean *every* observation in the ESA-Hubble archive?  That sounds like it might take a while!

jespinosaar replied:
We also wanted to cover the "no criteria" case to avoid failures in the algorithm in case no criteria is defined. If it is executed synchronously, it will only retrieve the first 2000 results. If it is executed asynchronously... yes, it might take a while!

keflavich commented:
You can ignore the codestyle errors that are in code you didn't change.  However, that might be an indication that you need to rebase against master - I hope we've fixed those, but I'm not sure.

jespinosaar commented:
Hello @keflavich, thanks for your input!
I have finally rebased against master, I think that was the best option not only for me, but for other branches in our fork. I have also included your comments and will push the updates soon.

jespinosaar commented:
Hello @keflavich, once we have solved the code style error, it seems there is another issue in the docs build but I cannot find the error, could you please tell me how to proceed?

keflavich commented:
The warnings are:
https://travis-ci.org/github/astropy/astroquery/jobs/696368365#L6443-L6450
which comes from having the wrong underlining in the docs.  I'll see if I can recommend a specific change inline

keflavich commented:
OK, on quick review, you have some bigger problems - it looks like your rebase against master somehow missed... something, I don't really understand what, but your changelog has an entry from NRAO in the diff, and it shouldn't.  Could you verify that you rebased against the _latest_ master?

Regarding the docs warning: that is in generated code that is not set here.  I don't understand where it's coming from.  @bsipocz , @ceb8, any ideas?  The problem is:
```
/home/travis/build/astropy/astroquery/docs/esa/hubble.rst:328: WARNING: Title level inconsistent:
astroquery.esa.hubble Package
-----------------------------
/home/travis/build/astropy/astroquery/docs/esa/hubble.rst:333: WARNING: Title level inconsistent:
Classes
^^^^^^^
```
which is in the apidoc generated code.  My best guess is that there's some formatting error somewhere further up, but it's hard to spot without rendering the docs.

jespinosaar commented:
Ok, thanks for your help! I have rebased against what I think is the last master, the one that has merged the branch 'e-koch-fix_nrao_projectcode_query' two days ago. Is that correct?

keflavich commented:
Yes, that is correct.  I don't understand why your CHANGES has the entry from that PR in it, though, in the diff - that shouldn't be there, because it is in master.

jespinosaar commented:
Yes, it should not be there... 

This is what I have done. In ESDC, we have a fork from astroquery repository (https://github.com/esdc-esac-esa-int/astroquery) that was not updated. Using this fork we created some branches (this one is called hubble_metadata_improvements), so they were not updated too. What I have rebased is our fork against astroquery master and, after that, I have rebased this branch against our updated fork.

keflavich commented:
Maybe that last step is the problem.  You should synchronize your `esdc-esac-esa-int/master` branch w/`astropy/master` if you're going to rebase against that, but generally it makes more sense to just rebase on `astropy/master`.

bsipocz commented on docs/esa/hubble.rst:
```diff
@@ -25,14 +25,68 @@ data, all HST data in the EHST are identical to those in MAST.
+  >>> help(esahubble)
+  Help on ESAHubbleClass in module astroquery.esa.hubble.core object:
+
+  class ESAHubbleClass(astroquery.query.BaseQuery)
```
`help` is a standard python command, there is no need to hardwire this in the docs, instead do link to the API doc, rather than doing this copy paste here.

ceb8 commented:
I would suggest the new function being renamed from `query_by_criteria` to `query_criteria` for consistency with other modules that have a criteria query (gemeni, simbad, mast).

jespinosaar commented:
Ok, I have rebased again following your instructions and solving the conflicts. I have also implemented your latest requests (query_criteria, help command removed and title style). Thanks for your help and feedback!

jespinosaar commented:
Ok, after implementing the changes another issue has arised in the tests,
`/home/travis/miniconda/envs/test/lib/python3.7/site-packages/astropy/utils/data.py:237: FileNotFoundError

=========================== short test summary info ============================

FAILED astroquery/utils/tests/test_utils.py::test_filecontainer_save - FileNo...

FAILED astroquery/utils/tests/test_utils.py::test_filecontainer_get - FileNot...

============ 2 failed, 831 passed, 522 skipped, 2 xfailed in 48.05s ============`

ceb8 commented:
I rebased this PR. @jespinosaar, please reset your local repository to match the github branch before adding any more commits.
```
git fetch origin
git reset --hard origin/hubble_metadata_improvements
```

jespinosaar commented:
Ok, done, thanks @ceb8! But it seems that the errors in the tests are still there...
`/home/travis/miniconda/envs/test/lib/python3.7/site-packages/astropy/utils/data.py:237: FileNotFoundError

=========================== short test summary info ============================

FAILED astroquery/utils/tests/test_utils.py::test_filecontainer_save - FileNo...

FAILED astroquery/utils/tests/test_utils.py::test_filecontainer_get - FileNot...

============ 2 failed, 831 passed, 522 skipped, 2 xfailed in 48.05s ============`

What should I do?

jespinosaar commented:
Hi @keflavich , @ceb8, I have already implemented the changes you requested, but I still get some errors in tests I have not modified, how should I proceed?

ceb8 commented on astroquery/esa/hubble/tests/test_esa_hubble.py:
```diff
@@ -32,32 +41,35 @@ def get_dummy_tap_handler(self):
+        ehst = ESAHubbleClass(self.get_dummy_tap_handler())
+        target_file = data_path('download_product')
+        ehst._request = MagicMock(return_value=target_file)
+        shutil.move = MagicMock(return_value=True)
```
This line is why you are getting the test failures. Whatever you moved here needs to be moved back when you are done. Is there a reason you use the `MagicMock` class rather than astroquery's `MockResponse`?

ceb8 commented:
@jespinosaar Looks like the tests are passing now, please add a changelog entry in the 4.2 section.

jespinosaar commented:
> 
> 
> @jespinosaar Looks like the tests are passing now, please add a changelog entry in the 4.2 section.

Yes, it seems that everything went well @ceb8  hehe. I have updated the changelog.

pep8speaks commented:
Hello @jespinosaar! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`astroquery/esa/hubble/core.py`](https://github.com/astropy/astroquery/blob/fc631f21a9074a532fa6201dc082ee04ef9b7836/astroquery/esa/hubble/core.py):

> [Line 40:1](https://github.com/astropy/astroquery/blob/fc631f21a9074a532fa6201dc082ee04ef9b7836/astroquery/esa/hubble/core.py#L40): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

ceb8 commented:
Thanks @jespinosaar.

jespinosaar commented:
Thanks @ceb8 , @keflavich , @bsipocz for your support!
